### PR TITLE
Adding Qt6 support

### PR DESCRIPTION
**Pull request by sandman7920**
This PR adds support for Qt6 and a new CMake configuration.
Both Qt 5 and 6 are supported.
The lowest-tested Qt5 version is 5.12.2 on Windows.
Qt5 is enabled by default, to build with Qt6 run cmake with **-DQT_MAJOR=Qt6**

Windows deployment is more automated:
```
cmake -GNinja -S d:\sqlitebrowser -B d:\sqlitebrowser_builds\release -DCMAKE_PREFIX_PATH="c:\dev\x64;C:\Qt\5.15.2\msvc2019_64" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=d:\sqlitebrowser_builds\DB.Browser.for.SQLite-x.xx-win64
cmake --build d:\sqlitebrowser_builds\release
cmake --install d:\sqlitebrowser_builds\release
```
This will build and deploy the project with all dependencies.

Currently, Qt6 depends on **Qt6::Core5Compat**, in the future, this should be dropped.

### CODE REVIEW

**Comment by lucydodo** [comment]
Before I dive in, it's probably a good idea to get CI working first.
The reason CI is failing right now is that it can't find the SQLCipher, SQLite, and Qt installations, and it used to hardlink the paths, but that doesn't seem to be the case now. (Of course, I think this is a better design)
Can you advise me on how to provide these paths as direct command arguemnts?
e.g.: `cmake -G Ninja -DSQLCIPHER_PREFIX="" -DSQLITE_PREFIX="" -DQT_PREFIX=""`

File: CMakeLists-legacy.txt
Comment (lucydodo) [inline CMakeLists-legacy.txt:1]: We can look at the commit history if we need to, so I think we'll be fine without this file.

Comment (sandman7920) [reply CMakeLists-legacy.txt:1]: On Windows, it should be

SQLite
```
-DCMAKE_PREFIX_PATH="C:/dev/OpenSSL-Win64;C:/dev/SQLite-Win64;C:/dev/Qt/5.12.12/msvc2017_64"
```

SQLCipher
```
-DCMAKE_PREFIX_PATH="C:/dev/OpenSSL-Win64;C:/git_repos/SQLCipher-Win64;C:/dev/Qt/5.12.12/msvc2017_64"
```

Those are paths set in the old configuration.
And same for 32bit

**Comment by sandman7920** [comment]
> Before I dive in, it's probably a good idea to get CI working first. The reason CI is failing right now is that it can't find the SQLCipher, SQLite, and Qt installations, and it used to hardlink the paths, but that doesn't seem to be the case now. (Of course, I think this is a better design) Can you advise me on how to provide these paths as direct command arguemnts? e.g.: `cmake -G Ninja -DSQLCIPHER_PREFIX="" -DSQLITE_PREFIX="" -DQT_PREFIX=""`

On Windows, it should be 

SQLite
```
-DCMAKE_PREFIX_PATH="C:/dev/OpenSSL-Win64;C:/dev/SQLite-Win64;C:/dev/Qt/5.12.12/msvc2017_64"
```

SQLCipher
```
-DCMAKE_PREFIX_PATH="C:/dev/OpenSSL-Win64;C:/git_repos/SQLCipher-Win64;C:/dev/Qt/5.12.12/msvc2017_64"
```

And same for 32bit

File: config/translations.cmake
Comment (mgrojo) [inline config/translations.cmake:22]: Is this having advantage of this integration: https://doc.qt.io/qt-6/cmake-commands-qtlinguisttools.html ? See https://github.com/sqlitebrowser/sqlitebrowser/discussions/3451#discussioncomment-6937104

By the way, which is the order of this list? Maybe chronological, and it would make sense to change it to an alphabetical sorting.

File: libs/qcustomplot-source/GPL.txt
Comment (mgrojo) [inline libs/qcustomplot-source/GPL.txt:1]: Why this change?

Comment (mgrojo) [reply libs/qcustomplot-source/GPL.txt:1]: Ok, if it comes upstream.

File: libs/qscintilla_2.14.1/ChangeLog
Comment (mgrojo) [inline libs/qscintilla_2.14.1/ChangeLog:1]: Do we need two qscintilla versions? Even if this version is for Qt 6 and the old one is for Qt5, it shouldn't come with the version number. Maybe a `qscintilla_qt6` path is needed.

File: src/ForeignKeyEditorDelegate.cpp
Comment (mgrojo) [inline src/ForeignKeyEditorDelegate.cpp:32]: In my opinion, if this is not really needed, it should be deleted. Otherwise, we need a comment about what is missing and when we can recover this code.

File: src/RemoteNetwork.cpp
Comment (mgrojo) [inline src/RemoteNetwork.cpp:16]: Why this? It seems incoherent with QRegExp being replaced by QRegularExpression for porting to Qt6.

**Review by mgrojo — COMMENTED** [review]
It compiles and starts on my usual development platform (Ubuntu 22, x86_64, Qt  5.15.3).

I'm no expert on CMake or the migration to Qt6, but I've dropped some comments about things to consider.

File: src/RemoteNetwork.cpp
Comment (sandman7920) [reply src/RemoteNetwork.cpp:16]: **QRegularExpression** does not supports **indexIn** and I did not want to rewrite
```cpp
   QRegExp regex("^.*modification-date=\"(.+)\";.*$");
            regex.setMinimal(true); // Set to non-greedy matching
            if(regex.indexIn(content_disposition) != -1)
                last_modified = QDateTime::fromString(regex.cap(1), Qt::ISODate);
```

For Qt6 QRegExp is provided by **Core5Compat** and it's not included by default

File: libs/qcustomplot-source/GPL.txt
Comment (sandman7920) [reply libs/qcustomplot-source/GPL.txt:1]: Yes, it comes from upstream.
https://www.qcustomplot.com/index.php/download

File: config/translations.cmake
Comment (sandman7920) [reply config/translations.cmake:22]: I don't know the minimum Qt5 version we need to support. I have compiled it with Qt 5.12.2 on Windows.
The order is taken directly from the old config.

File: libs/qscintilla_2.14.1/ChangeLog
Comment (mgrojo) [reply libs/qscintilla_2.14.1/ChangeLog:1]: So you plan to move `libs/qscintilla_2.14.1` over `libs/qscintilla` ?

File: config/translations.cmake
Comment (sandman7920) [reply config/translations.cmake:16]: The list is sorted

File: src/RemoteNetwork.cpp
Comment (mgrojo) [reply src/RemoteNetwork.cpp:16]: Is this of any help? 
> replace if (rx.indexIn( by QRegularExpressionMatch match = rx.match(searchString); if (match.hasMatch()) and update the invalid code parts

From https://github.com/Komet/MediaElch/issues/1086

Comment (sandman7920) [reply src/RemoteNetwork.cpp:16]: I want to drop **Core5Compat**but, there are some edge cases with **QRegExp** and **TextCodecs** and I am not confident enough to rewrite it (especially regex see CrowdStrike :) ).

Comment (mgrojo) [reply src/RemoteNetwork.cpp:16]: Ok, if Core5Compat is working, no problem on my part. For me, it could be merged when you consider it ready.

**Review by mgrojo — APPROVED** [review]
For me, the changes are acceptable. It can be merged when the checks are passing.

**Comment by justinclift** [comment]
@sandman7920 Uh oh, it looks like you merged without squashing the (many) commits into a single one.

Are you up for fixing the commit history on `master`, as it's a bit of a mess now. :wink:

Note that you're not the first person to do that before, I think we've all done it at one time or another. :innocent:

**Comment by sandman7920** [comment]
@justinclift Sorry, can we do something?

**Comment by justinclift** [comment]
Yeah, it'll need to be fixed manually in the `master` branch.

i.e. do a manual squash of the commits and push the fixed history

Are you ok to do that?  If you're a it unsure, just let me know and I'll get it done. :smile:

**Comment by sandman7920** [comment]
@justinclift Better do it, my git/github skills are mediocre at best.

**Comment by justinclift** [comment]
No worries, I'll get it done.  As as general guide, these are the steps I'll be taking in my local repo:

### Make sure my repo is clean and doesn't have leftover stuff from other 

```
$ git clean -dffx
$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

### Ensure I have the latest source

```
$ git pull
```

### Find the last git commit that we want to rewind everything to

```
$ git log
[do some manual paging]
commit 57f1824df1f02040ceac404a07b29cfbbcb2b45f
Author: Nikolay Zlatev <sandman7920@gmail.com>
Date:   Fri Aug 16 16:15:28 2024 +0300

    Optimize build process with Qt6 in mind.

commit f2203887dff9536148f91613093b0f27bea29f7c
Author: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Date:   Tue Aug 13 20:07:07 2024 +0900

    build(deps): bump signpath/github-action-submit-signing-request from 0.4 to 1 (#3704)
    
    Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
```

So, `f2203887dff9536148f91613093b0f27bea29f7c` is the last commit before the merge-that-wasn't-squashed.


### Rewind to that commit

```
$ git reset f2203887dff9536148f91613093b0f27bea29f7c
```

The repository should now be at the at the "last" commit before the merge, but still have all of your PR changes applied to it.

```
$ git log
commit f2203887dff9536148f91613093b0f27bea29f7c (HEAD -> master)
Author: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Date:   Tue Aug 13 20:07:07 2024 +0900

    build(deps): bump signpath/github-action-submit-signing-request from 0.4 to 1 (#3704)
    
    Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
```

and:

```
$ git status
On branch master
Your branch is behind 'origin/master' by 59 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   .github/workflows/build-macos.yml
        modified:   .github/workflows/build-windows.yml
[etc]
```

At this point, I'll use `git add` and similar to make a single commit from all of your changes, then push it to `master` using your details as the commit author.

Does that make sense? :smile:

**Comment by justinclift** [comment]
Hmm, actually that might not be the right commit to wind back to.  I think that would get some earlier ones as well.

Doing some more investigation...

**Comment by justinclift** [comment]
Ahhh.  Nah, that's the right commit it just has @lucydodo's one to update the `currentrelease`  thrown into the middle of things as well.  Should still be workable.

**Comment by justinclift** [comment]
Ok, done.  At least, I *think* I got it right.  Lets see if the CI passes, just in case I missed something somehow. :wink:

**Comment by justinclift** [comment]
Hmm, CI is failing so it looks like I missed *something*.  Any ideas what I missed, as I'm not seeing anything obvious?

**Comment by sandman7920** [comment]
After the merge, I pushed build/signing code to the master branch.

**Comment by sandman7920** [comment]
I will pull and check the diff

**Comment by justinclift** [comment]
> After the merge, I pushed build/signing code to the master branch.

Ugh, I totally forgot about that.  So those commits became part of my merged "Adding Qt6 support" squashed patch.

**Comment by justinclift** [comment]
With the CI errors happening at the moment, the errors are saying:

```
CMake Error at CMakeLists.txt:11 (include):
  include could not find requested file:

    config/options.cmake
```

I have those files on my system though:

```
$ ls -al config/
total 13
drwxr-xr-x  2 jc jc    9 Aug 26 21:12 ./
drwxr-xr-x 13 jc jc   25 Aug 26 21:27 ../
-rw-r--r--  1 jc jc  455 Aug 26 21:12 3dparty.cmake
-rw-r--r--  1 jc jc 2581 Aug 26 21:12 install.cmake
-rw-r--r--  1 jc jc  836 Aug 26 21:12 options.cmake
-rw-r--r--  1 jc jc 1125 Aug 26 21:12 platform_apple.cmake
-rw-r--r--  1 jc jc  603 Aug 26 21:12 platform.cmake
-rw-r--r--  1 jc jc 1802 Aug 26 21:12 platform_win.cmake
-rw-r--r--  1 jc jc 1504 Aug 26 21:12 translations.cmake
```

**Comment by sandman7920** [comment]
Yes, *.cmake is added to the ignore list by default (maybe not to be added by mistake?) and I have added them whit -f

**Comment by sandman7920** [comment]
Should I push them back?

**Comment by justinclift** [comment]
Ahhh.  Nah, I'll add them back to the source tree as part of the squashed commit.  1 sec...

**Comment by justinclift** [comment]
k, just force pushed an updated, squashed `Adding Qt6 support` commit.  Hopefully it's 100% good now. :smile:

**Comment by justinclift** [comment]
k, CI is passing with the new, squashed `Adding Qt6 support` commit.  If you have some time, it's probably a good idea to look it over to make sure there's nothing else I missed.  Just in case.

Oh, and your follow up signing commits... well, lets just leave them merged into this one for simplicity. :grin:

**Comment by sandman7920** [comment]
Everything seems fine.
But the issue with `qsvgiconengine` plugin not being shipped remains, I am not familiar with `WiX Toolset` at all.

**Comment by justinclift** [comment]
Ahhh.  @lucydodo might have some ideas about that then. :smile:

**Comment by justinclift** [comment]
It *might* just be a matter of us needing to include some additional .dll files in the .xml definition used to build the package.

**Comment by sandman7920** [comment]
Yes, it is but in the future, it will be better for `WiX Tools` to include the directory structure generated by the `cmake --install` step (not executed at the moment).
`cmake install` uses `windeployqt` which inspects the DB4S executable and deploys all necessary runtime libraries.

**Comment by lucydodo** [comment]
@sandman7920 May I ask where the `qsvgiconengine` file you uploaded last time came from?
I configured a Windows VM, and built DB4S, but couldn't find that library file. :(

**Comment by sandman7920** [comment]
It should be in QT_DIR / "plugins/iconengines/qsvgicon.dll"

**Comment by lucydodo** [comment]
Thanks for letting me know, I've changed the code to include that plugin and are testing it now.

**Comment by lucydodo** [comment]
Resolved via 2c50024435b510e70f89e7ff38c46c543b8533b5, @sandman7920 Thank you for your advice and assistance.

**Pull request merged**
